truffle/com.oracle.truffle.api.object.dsl/src/com/oracle/truffle/api/object/dsl/Layout.java

```diff
@@ -0,0 +1,299 @@
+ * object contains that is used to look up shape properties) is created by the layout and then used
+ * when creating new instances. As shape properties are set and changed, multiple factories will be
+ * created and it is up to the user to store and supply these as needed.
+ *
+ * Consider the example of a Java-style object, with a class and a hash code. The class would be a
+ * shape property, as many objects will share the same class, and the hash code will be a normal
+ * property.
+ *
+ * Shape properties are created by parameters in the method that creates the shape. The factory
+ * method then accepts an instance of a factory when creating the object, which is how the instance
+ * knows the value of the class property to use. A getter for a shape property can be defined as
+ * normal.
+ *
+ * <pre>
+ * &#64;Layout
+ * interface JavaObjectLayout {
+ *
+ *     DynamicObjectFactory createJavaObjectShape(JavaClass klass);
+ *
+ *     DynamicObject createJavaObject(DynamicObjectFactory factory, int hashCode);
+ *
+ *     JavaClass getKlass(DynamicObject object);
+ *
+ *     int getHashCode(DynamicObject object);
+ *
+ * }
+ * </pre>
+ *
+ * <pre>
+ * JavaClass getKlass(DynamicObjectFactory factory);
+ *
+ * JavaClass getKlass(ObjectType objectType);
+ * </pre>
+ *
+ * When we load our Java interpreter we need to set the class property of the {@code Class} object
+ * to be itself. This means in this one isolated, slow-path, case we need to change a shape property
+ * for an object that is already allocated. Getters for shape properties can be defined for the
+ * {@link DynamicObjectFactory}, and for the {@link ObjectType}.
+ *
+ * Setters for shape properties are more complex, and they are not intended to be used in the fast
+ * path. Setters can be defined on a {@link DynamicObjectFactory}, in which case they return a new
+ * factory, or on a {@link DynamicObject}, in which they they change the shape of the object. This
+ * is a slow-path operation and is likely to invalidate caches in your interpreter.
+ *
+ * <pre>
+ * DynamicObjectFactory setKlass(DynamicObjectFactory factory, JavaClass value);
+ *
+ * void setKlass(DynamicObject object, JavaClass value);
+ * </pre>
+ *
+ * Apply this to our example with Java classes:
+ *
+ * <pre>
+ * javaClassObject = JavaObjectImpl.INSTANCE.createJavaObject(JavaObjectImpl.INSTANCE.createJavaObjectShape(null), defaultHashCode());
+ * JavaObjectImpl.INSTANCE.setKlass(javaClassObject, javaClassObject);
+ * </pre>
+ *
+ * <h2>Layout Inheritance</h2>
+ *
+ * Inheritance of layout interfaces allows you to model classical class inheritance, such as in a
+ * language like Java. Use normal interface inheritance to make one layout inherit from another. You
+ * then need to add the parameters for super-layouts at the beginning of sub-layout constructor
+ * methods.
+ *
+ * Inherited shape properties work in a similar way.
+ *
+ * <pre>
+ * &#64;Layout
+ * interface BaseLayout {
```

Comment:
Following comment depends on result of #19: I want all our code samples to be provided via the code snippets infrastructure - e.g. to be error proof, hyperlinked and not printed in a hard to read uniform font/color.